File: app/src/main/kotlin/com/ghissue/app/storage/IssueQueueManager.kt
package com.ghissue.app.storage

import android.content.Context
import android.content.SharedPreferences
import com.ghissue.app.network.ApiClient
import com.ghissue.app.network.CreateIssueRequest
import com.ghissue.app.network.QueuedIssue
import com.squareup.moshi.Types
import java.io.IOException

class IssueQueueManager(context: Context) {

    private val prefs: SharedPreferences =
        context.getSharedPreferences("ghissue_queue", Context.MODE_PRIVATE)

    private val adapter = ApiClient.moshi.adapter<List<QueuedIssue>>(
        Types.newParameterizedType(List::class.java, QueuedIssue::class.java)
    )

    @Synchronized
    fun enqueue(issue: QueuedIssue) {
        val list = getAll().toMutableList()
        list.add(issue)
        prefs.edit().putString(KEY_QUEUED_ISSUES, adapter.toJson(list)).apply()
    }

    @Synchronized
    fun remove(id: String) {
        val list = getAll().filter { it.id != id }
        prefs.edit().putString(KEY_QUEUED_ISSUES, adapter.toJson(list)).apply()
    }

    @Synchronized
    fun getAll(): List<QueuedIssue> {
        val json = prefs.getString(KEY_QUEUED_ISSUES, null) ?: return emptyList()
        return adapter.fromJson(json) ?: emptyList()
    }

    fun count(): Int = getAll().size

    suspend fun drainQueue(tokenStore: TokenStore): DrainResult {
        var submitted = 0
        var failed = 0
        var stopReason: String? = null

        val queue = getAll()
        for (issue in queue) {
            try {
                val token = "Bearer ${tokenStore.accessToken}"
                ApiClient.gitHubApi.createIssue(
                    token = token,
                    owner = issue.owner,
                    repo = issue.repo,
                    request = CreateIssueRequest(
                        title = issue.title,
                        body = issue.body,
                        labels = issue.labels
                    )
                )
                remove(issue.id)
                submitted++
            } catch (e: IOException) {
                stopReason = "network"
                failed++
                break
            } catch (e: retrofit2.HttpException) {
                when (e.code()) {
                    401 -> {
                        stopReason = "auth"
                        failed++
                        break
                    }
                    else -> {
                        remove(issue.id)
                        failed++
                    }
                }
            }
        }

        return DrainResult(submitted, failed, stopReason)
    }

    data class DrainResult(
        val submitted: Int,
        val failed: Int,
        val stopReason: String?
    )

    companion object {
        private const val KEY_QUEUED_ISSUES = "queued_issues"
    }
}

File: ios/GhIssue/Storage/IssueQueueManager.swift
import Foundation

actor IssueQueueManager {
    private let defaults = UserDefaults(suiteName: Constants.appGroupID) ?? .standard
    private let queueKey = "queued_issues"

    struct DrainResult {
        let submitted: Int
        let failed: Int
        let stopReason: String?
    }

    func enqueue(_ issue: QueuedIssue) {
        var list = getAll()
        list.append(issue)
        save(list)
    }

    func remove(id: String) {
        let list = getAll().filter { $0.id != id }
        save(list)
    }

    func getAll() -> [QueuedIssue] {
        guard let data = defaults.data(forKey: queueKey) else { return [] }
        return (try? JSONDecoder().decode([QueuedIssue].self, from: data)) ?? []
    }

    func count() -> Int {
        getAll().count
    }

    func drainQueue() async -> DrainResult {
        guard let token = KeychainStore.accessToken else {
            return DrainResult(submitted: 0, failed: 0, stopReason: "auth")
        }

        var submitted = 0
        var failed = 0
        var stopReason: String?

        let queue = getAll()
        for issue in queue {
            do {
                _ = try await GitHubAPIClient.createIssue(
                    owner: issue.owner,
                    repo: issue.repo,
                    request: CreateIssueRequest(
                        title: issue.title,
                        body: issue.body,
                        labels: issue.labels
                    ),
                    token: token
                )
                remove(id: issue.id)
                submitted += 1
            } catch let error as GitHubAPIError {
                switch error {
                case .networkError:
                    stopReason = "network"
                    failed += 1
                    return DrainResult(submitted: submitted, failed: failed, stopReason: stopReason)
                case .unauthorized:
                    stopReason = "auth"
                    failed += 1
                    return DrainResult(submitted: submitted, failed: failed, stopReason: stopReason)
                case .httpError:
                    remove(id: issue.id)
                    failed += 1
                }
            } catch {
                stopReason = "network"
                failed += 1
                return DrainResult(submitted: submitted, failed: failed, stopReason: stopReason)
            }
        }

        return DrainResult(submitted: submitted, failed: failed, stopReason: stopReason)
    }

    // MARK: - Private

    private func save(_ issues: [QueuedIssue]) {
        let data = try? JSONEncoder().encode(issues)
        defaults.set(data, forKey: queueKey)
    }
}
